core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -2510,7 +2510,7 @@ private RelBuilder aggregate_(GroupKeyImpl groupKey,
     // duplicates, then add a Project.
     final Set<AggregateCall> callSet = new HashSet<>();
     final PairList<Integer, @Nullable String> projects = PairList.of();
-    Util.range(groupSet.cardinality())
+    Util.range(groupSet2.cardinality())
```

Comment:
when is groupSet.cardinality() != groupset2.cardinality()?

Reply:
From my understanding the return value of `groupSet.cardinality()` should always be the same as for `groupSet2.cardinality()` in the current implementation. My main point is that with all the logic at the beginning of the `aggregate_()` method which introduces `groupSet2` we should use it consistently in the latter part of the method.

Reply:
Sure, but this means that #3700 is actually equivalent to this fix.
So that PR probably fixes both issues as well.
Is there a more descriptive name that could be used instead of groupSet2? Maybe that's the root problem: it is very hard to tell what groupSet2 means. But if it was called e.g., replacementGroupSet, then it would become apparent where it is misused.

Reply:
I have approved the PR, but if you are willing to improve the code I will wait for a new commit.

Reply:
I can give it a shot on Monday when I'm back in the office.

Reply:
- I renamed `groupSet2` and `groupSets2` to `groupSetAfterPruning` and `groupSetsAfterPruning` respectively to give them more obvious names.
- I also renamed the helper variable `map` to `sourceFieldToTargetFieldMap` to make it more obvious.
- I moved the field pruning and agg call deduplication code into its own `pruneAggregateInputFieldsAndDeduplicateAggCalls ` method to reduce the length of the original `aggregate_()` method.